### DISCUSSION TITLE
Creating even more 'Powerwall state of charge' options for automations

### DISCUSSION THREAD

**Discussion by nickz98** [netzero-labs/netzero#17] [Ideas]
Hi, I note the 'Powerwall state of charge' automation options.

In the UK (Worcester) it's not a good solar day, just too cloudy, and I'm only likely to generate 50% of the electricity I did a week ago for example.  Which means I'm looking for more control over the powerwall, to help tell it when to charge from the grid. 
 i.e. I know my solar generation will be low, my household usage will remain approximately the same /  be slightly higher due to the poor weather.

So could we get options something like if:
a) the powerwall charge level is below x % 
b) for today / the next 24hrs we can see when the cheapest times are (1/2 hour chunks from Octopus agile contract).
then
tell / encourage the powerwall to charge from the grid during those cheapest times (perhaps we can say the 'x' (e.g. 5) cheapest 1/2 hrs).  Perhaps it could temporarily set the relevant 1/2 hours to be TBC + 'off-peak' which is how I used to manually control it before your great app.

For example today 17th May 2024 - I can see the cheapest times now to charge from the grid are between 1:30pm and 3:30pm - I know in those 2 hours I can charge my powerwall 2 approx 66% (33% per hour) my current level of charge at 9am is 1%, and I'm only currently charging at 1.2kw - suggesting unless the weather improves my charge level by 4pm otherwise will be quite low,  I know that TBC should in theory clock all this and self charge from the grid....but I'm quite confident it wont.

Even better if we're able to see what the predicted generation for today is estimated to be, as the cheapest times on agile at the moment are in the early hours and the lead up to 4pm. & 10:30pm to 11:00pm.  So we could add to the 'if' conditions when the predicted generation is expected to be below 'x' over the next x hours etc.   Getting more and more complicated I guess to work out the rules we'd want to set .   But for not TBC or self powered alone don't seem to cope as well as they could for cloudy days.

Thoughts?

**Comment by nickz98** [comment]
Adding to the above.
TBF the weather picked up a little and by 2/ 2:30 pm the charge was up to 80%, and I noticed at 1 point the powerwall took some charge form the grid at a cheap time - I hope / think to assist it getting to 100% by 4pm (it's been set to TBC all day), then I flip it to SP at 4pm as a precaution against any charge from the grid for a few hours at the expensive time.

So on this occasion I would have set my setting I'm suggesting above such they they wouldn't have actually kicked in.  But on another day I think they would be needed.

**Comment by zigam** [comment]
Thanks for the suggestion!  As you noted, these rules could get pretty complicated.  Tesla's Time-Based Control mode is supposed to take into account all the necessary variables: electricity prices, predicted solar generation and house usage.  But of course they're predictions, so they might not always do the optimal thing.

I need to think more about ways Netzero can facilitate TBC, without too much manual intervention.
